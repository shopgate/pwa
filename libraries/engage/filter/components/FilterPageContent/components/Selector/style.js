import { css } from 'glamor';

export const accordion = css({
  overflow: 'hidden',
}).toString();

export const content = css({
  marginLeft: -8,
  marginBottom: -8,
});
