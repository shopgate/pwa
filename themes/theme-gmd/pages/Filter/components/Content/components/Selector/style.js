import { css } from 'glamor';

export const content = css({
  marginLeft: -8,
  marginBottom: -8,
});
