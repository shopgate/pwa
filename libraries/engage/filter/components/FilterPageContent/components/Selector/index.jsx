import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Accordion, SurroundPortals } from '@shopgate/engage/components';
import { i18n } from '@shopgate/engage/core';
import { FilterItem } from '@shopgate/engage/filter';
import { PORTAL_FILTER_SELECTOR } from '@shopgate/engage/filter/constants';
import ValueButton from './components/ValueButton';
import Toggle from './components/Toggle';
import Selected from './components/Selected';
import * as styles from './style';

/**
 * The selector component.
 */
class Selector extends PureComponent {
  static propTypes = {
    id: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    values: PropTypes.arrayOf(PropTypes.shape()).isRequired,
    multi: PropTypes.bool,
    onChange: PropTypes.func,
    selected: PropTypes.node,
  };

  static defaultProps = {
    multi: false,
    onChange() { },
    selected: null,
  };

  state = {
    selected: this.props.selected || [],
  };

  /**
   * @param {Object} nextProps The new incoming props.
   */
  UNSAFE_componentWillReceiveProps({ selected }) {
    if (selected !== this.state.selected) {
      this.setState({ selected });
    }
  }

  /**
   * @param {SyntheticEvent} event The button click event.
   */
  handleClick = (event) => {
    const { value } = event.target;
    const { selected } = this.state;
    const { id, multi, onChange } = this.props;
    let newSelected = [...selected, value];

    // If in single select mode, only allow one selected value.
    if (!multi && selected.length === 1) {
      newSelected = [value];
    }

    // If the clicked value was already selected, remove it again.
    if (selected.includes(value)) {
      newSelected = selected.filter(item => item !== value);
    }

    // Set it if it wasn't selected already.
    this.setState({ selected: newSelected });
    onChange(id, newSelected);
  }

  /**
   * Filter value change handler for the portal props. Invokes handleClick method with a mocked
   * click event
   * @param {string} updatedId Id of the updated filter value
   */
  handlePortalChange = (updatedId) => {
    this.handleClick({ target: { value: updatedId } });
  }

  /**
   * @param {Object} props The send render props.
   * @return {JSX}
   */
  renderLabel = (props) => {
    const { label, values } = this.props;
    const { selected } = this.state;

    return (
      <Toggle
        {...props}
        label={label}
        selected={<Selected values={values} selected={selected} />}
      />
    );
  }

  /**
   * @returns {JSX}
   */
  render() {
    const {
      values, id, label, multi,
    } = this.props;
    const { selected } = this.state;

    return (
      <SurroundPortals
        portalName={PORTAL_FILTER_SELECTOR}
        portalProps={{
          filter: {
            id,
            label,
            values,
            isMultiSelect: multi,
          },
          selectedValueIds: selected,
          onChange: this.handlePortalChange,
        }}
      >
        <FilterItem>
          <Accordion
            renderLabel={this.renderLabel}
            testId={id}
            handleLabel={i18n.text('filter.filter_by', { label })}
            className={styles.accordion}
          >
            <div className={styles.content}>
              {values.map(value => (
                <ValueButton
                  key={value.id}
                  id={value.id}
                  label={value.label}
                  isActive={(selected && selected.includes(value.id))}
                  onClick={this.handleClick}
                />
              ))}
            </div>
          </Accordion>
        </FilterItem>
      </SurroundPortals>
    );
  }
}

export default Selector;
