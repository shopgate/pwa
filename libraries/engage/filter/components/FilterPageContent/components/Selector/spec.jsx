import React from 'react';
import { shallow } from 'enzyme';
import Selector from './index';

jest.mock('@shopgate/engage/core', () => ({
  hasWebBridge: () => false,
  i18n: {
    text: string => string,
  },
}));
jest.mock('@shopgate/engage/components', () => ({
  Accordion: ({ children }) => children,
}));
jest.mock('@shopgate/engage/filter', () => ({
  FilterItem: ({ children }) => children,
}));

const values = [
  {
    id: 'foo',
    label: 'fooLabel',
  },
  {
    id: 'bar',
    label: 'barLabel',
  },
];

describe('Filter: <Selector />', () => {
  it('should render with minimum props', () => {
    const wrapper = shallow(<Selector id="foo" label="bar" values={values} />);
    expect(wrapper).toMatchSnapshot();
  });
});
