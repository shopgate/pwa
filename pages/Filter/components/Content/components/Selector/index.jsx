import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import { Accordion } from '@shopgate/engage/components';
import { i18n } from '@shopgate/engage/core';
import Item from '../Item';
import ValueButton from './components/ValueButton';
import Toggle from './components/Toggle';
import Selected from './components/Selected';
import * as styles from './style';

/**
 * The selector component.
 */
class Selector extends PureComponent {
  static propTypes = {
    id: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    values: PropTypes.arrayOf(PropTypes.shape()).isRequired,
    multi: PropTypes.bool,
    onChange: PropTypes.func,
    selected: PropTypes.node,
  };

  static defaultProps = {
    multi: false,
    onChange() { },
    selected: null,
  };

  state = {
    selected: this.props.selected || [],
  };

  /**
   * @param {Object} nextProps The new incoming props.
   */
  componentWillReceiveProps({ selected }) {
    if (selected !== this.state.selected) {
      this.setState({ selected });
    }
  }

  /**
   * @param {SyntheticEvent} event The button click event.
   */
  handleClick = (event) => {
    const { value } = event.target;
    const { selected } = this.state;
    const { id, multi, onChange } = this.props;
    let newSelected = [...selected, value];

    // If in single select mode, only allow one selected value.
    if (!multi && selected.length === 1) {
      newSelected = [value];
    }

    // If the clicked value was already selected, remove it again.
    if (selected.includes(value)) {
      newSelected = selected.filter(item => item !== value);
    }

    // Set it if it wasn't selected already.
    this.setState({ selected: newSelected });
    onChange(id, newSelected);
  }

  /**
   * @param {Object} props The send render props.
   * @return {JSX}
   */
  renderLabel = (props) => {
    const { label, values } = this.props;
    const { selected } = this.state;

    return (
      <Toggle
        {...props}
        label={label}
        selected={<Selected values={values} selected={selected} />}
      />
    );
  }

  /**
   * @returns {JSX}
   */
  render() {
    const { values, id, label } = this.props;
    const { selected } = this.state;

    return (
      <Item>
        <Accordion
          renderLabel={this.renderLabel}
          testId={id}
          handleLabel={i18n.text('filter.filter_by', { label })}
        >
          <div className={styles.content}>
            {values.map(value => (
              <ValueButton
                key={value.id}
                id={value.id}
                label={value.label}
                isActive={(selected && selected.includes(value.id))}
                onClick={this.handleClick}
              />
            ))}
          </div>
        </Accordion>
      </Item>
    );
  }
}

export default Selector;
