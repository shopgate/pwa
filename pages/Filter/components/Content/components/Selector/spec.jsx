import React from 'react';
import { shallow } from 'enzyme';
import Selector from './index';

const values = [
  {
    id: 'foo',
    label: 'fooLabel',
  },
  {
    id: 'bar',
    label: 'barLabel',
  },
];

describe('Filter: <Selector />', () => {
  it('should render with minimum props', () => {
    const wrapper = shallow(<Selector id="foo" label="bar" values={values} />);
    expect(wrapper).toMatchSnapshot();
  });
});
