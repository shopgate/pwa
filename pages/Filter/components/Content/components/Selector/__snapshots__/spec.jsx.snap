// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter: <Selector /> should render with minimum props 1`] = `
<FilterItem>
  <Accordion
    className={null}
    contentClassName={null}
    handleLabel="filter.filter_by"
    renderLabel={[Function]}
    testId="foo"
  >
    <div
      className={
        Object {
          "data-css-bb7tjd": "",
        }
      }
    >
      <ValueButton
        id="foo"
        isActive={false}
        key="foo"
        label="fooLabel"
        onClick={[Function]}
      />
      <ValueButton
        id="bar"
        isActive={false}
        key="bar"
        label="barLabel"
        onClick={[Function]}
      />
    </div>
  </Accordion>
</FilterItem>
`;
